include/swift/AST/DiagnosticsSema.def

```diff
@@ -1796,8 +1796,8 @@ ERROR(use_of_equal_instead_of_equality,none,
       "use of '=' in a boolean context, did you mean '=='?", ())
 
 ERROR(type_cannot_conform, none,
-      "%select{|value of protocol }0type %1 cannot conform to %2; "
-      "only struct/enum/class types can conform to protocols",
+      "%select{|A value of protocol }0type %1 does not conform to the protocol itself (exception: 'Error'); "
+      "only concrete types such as structs, enums and classes can conform to protocols",
```

Comment:
Three suggestions:

1) The diagnostics should always start with lowercase letter, and you don’t need “A” at the beginning here, since the diagnostics use an abbreviated form of English known as RWR (reduced written register) 🙂

2) I’m not sure if it’s worth special casing `Error` because we’d have to change the diagnostic again if Swift gains additional self-conforming protocols, but perhaps it’s okay for now.

3) The “type %1 does not conform to the protocol itself” part is only correct if both types are the same. I think you need two separate diagnostics here - one for a generic “does not conform” and one for self-conformance, refined conformance, etc.

Reply:
@theblixguy Thanks for your feedback, I will work on them.

Reply:
I’m not sure there’s anything to gain from splitting this into several diagnostics. Perhaps we would be better off just mentioning the contrary — `; protocol types cannot conform to other existential types`— to cover abstract conformances as well.

Reply:
Sure, a single diagnostic would be preferable (sometimes separate ones are easier to manage when the diagnostic text/select logic is more complicated, but it depends on what text we settle for this one).

Reply:
I don't think it's bad to have a _tailored_ diagnostic for the protocol type not conforming to the protocol itself - that's exactly what the SR is asking for, because reading something like `Value of protocol type 'Cancellable' cannot conform to 'Cancellable'` is confusing. We could definitely put this into a "single" diagnostic and select which text to use based on a bool like @theblixguy mentioned.

Reply:
I also recommend trying to find a way to phrase this error text in the same way you would a headline. Suppose you had code that was trying to use an existential in such a way, and try to complete the phrase "extra, extra, ____"

Reply:
Alternatively, that extra piece of information could be attached as a note. We're pretty inconsistent about when information is presented after `;` in the main error message and when it's attached as a note. In this case, I don't think it's particularly important to show that information in the main error message (but others may disagree!).

Reply:
> I don't think it's bad to have a _tailored_ diagnostic for the protocol type not conforming to the protocol itself - that's exactly what the SR is asking for, because reading something like `Value of protocol type 'Cancellable' cannot conform to 'Cancellable'` is confusing. We could definitely put this into a "single" diagnostic and select which text to use based on a bool like @theblixguy mentioned.

The self-conforming case is pretty rare, whereas `value of protocol type ‘Comparable’ does not conform to ‘Equatable’` is just as confusing unless you read the full error. I believe the extra info coming after `;` should be enough to clarify the error for all cases, and I’m worried we may end up raising more questions by unnecessarily sprinkling sugar over one specific case.

> I don't think it's particularly important to show that information in the main error message (but others may disagree!).

I would prefer to keep it in the main message for 2 reasons:
- It carries an important clarification, but Xcode doesn’t reveal attached notes inline unless you tap on the main error.
- I tend to treat notes more as representing hints or non-interactive ways to solve the error/warning being attached to.

Reply:
> The self-conforming case is pretty rare

I actually think this is a common misconception.

> I tend to see notes more as representing hints or non-interactive ways to resolve the error/warning they’re attached to.

I agree, and I think this note falls into that case. I think the note is hinting at passing an argument with a non-existential type, but that's not necessarily the "right" fix for the developer. For example, perhaps the developer actually wants to change whatever requires the conformance to take an existential instead.

> Xcode doesn’t reveal attached notes inline unless you tap on the main error

I definitely understand the discoverability issue with notes in Xcode, but I also don't think it's right to cram as much clarifying information as possible into the one-line error message to make up for that.

Is there a way we can clarify the main error message _and_ move the extra information about types that can conform into a note? I've been trying to avoid the term "existential" here - is that the clarifying piece that you think is missing from the first clause @AnthonyLatsis ?

Reply:
I'm starting to think that using the term "protocol type" to avoid "existential" is what makes this error message confusing.

Reply:
The terminology used in TSPL is "protocol type". The [Protocols page](https://docs.swift.org/swift-book/LanguageGuide/Protocols.html) has 4 mentions of "protocol type" and only one mention of "existential type". I agree with Brent's point in the SR that we should avoid using "existential" in the diagnostic if possible, but having it in the educational note would be fine IMO as we have more space to explain details.

Reply:
I know it's the terminology in TSPL, but in the longer form text there's more clarification about what it means to use a protocol as a type. TSPL also introduces the term "existential type". I'm not saying that we absolutely should use "existential type" here - I'm simply wondering whether saying "protocol type X cannot conform to protocol X" might be causing some confusion.

Reply:
What if we switched it around to say `protocol X used as a type cannot conform to protocol Y`. Is that any better? It seems to make it more clear that there's a distinction between a protocol and using a protocol as a type

Reply:
> What if we switched it around to say protocol X used as a type cannot conform to protocol Y or protocol X as a type cannot conform to protocol Y. Is that any better?

I like this option better. I agree that the similarity might create more confusion, especially for non-native speakers.

Reply:
“only struct/enum/class types can conform to protocols”

I am wondering if we should also mention tuples in the educational notes, because as of SE-0283, tuples conform to Equatable, Hashable & Comparable. What do you think?

Reply:
Since SE-0283 hasn't been merged yet, we can update the educational note as part of that #28833 or as a follow-up after both this PR and 28833 have been merged.